### ISSUE TITLE
Torch SpatialBatchNormalization layer cannot be loaded from weights

### ISSUE THREAD

**Issue by Liuftvafas** [NVIDIA/DIGITS#496]
During the training a snapshot of model is saved for later use. It consists of model script, weights snapshot, mean data, and labels file.

However, if model contains SpatialBatchNormalization layer that's not enough. The layer contains fields **running_mean** and **running_std** which are calculated from training data but later lost during save/load snapshot procedure.

Maybe saving a sanitized model (with cleared intermediate states) instead of weights would be a good idea?

**Comment by gheinrich** [comment]
Thanks for the report @Liuftvafas. Indeed we're using the top-level module's `getParameters()` method to save/reload learnable parameters.

 In my understanding, `running_mean` and `running_var` are learned during training. Together with `weight=gamma` and `bias=beta`, these comprise the set of learnable parameters in `nn.SpatialBatchNormalization`. Therefore, I feel it would make sense to add `running_mean` and `running_var` to the parameters that are returned by `getParameters()`. Possibly this could be implemented by overloading `nn.Module:getParameters()` in `nn.SpatialBatchNormalization`? Doing it this way would keep the model abstraction unchanged, right?

cc @colesbury for information as he authored a recent rework of `nn.SpatialBatchNormalization`. Sorry for the spam but I think you might be interested in contributing to this discussion. Thanks in advance!

Note: we're not using `torch.save` and `torch.load` to save/load a model since we're only interested in saving the model for FPROP and we don't want to save gradients, which consume a lot of disk space.

**Comment by Liuftvafas** [comment]
`running_mean` and `running_var` are not learned in the sense how weights and biases are learned via back propagation in training. They are rather calculated from training data. If that does not contradict the **learnable parameters** idea, then it would be a good idea to add them to `getParameters()`.

I was introduced with a model `sanitize()` function in @soumith imagenet-multiGPU.torch code which removes gradients to save disk space but does not remove parameters like `running_*`. Model takes only 2x of weights size on disk after it is applied.

**Comment by gheinrich** [comment]
Thanks, now I see what you mean by **sanitized** model.  I'm interested to know what you all think about adding `running_mean` and `running_var` to the table returned by `parameters()` and if it's better or worse than a `sanitize()` function.

**Comment by colesbury** [comment]
I think adding `running_mean` and `running_var` to `parameters()` will break the optim package, since these parameters are *not* learned via SGD (or the other optimization methods in optim).

I think the best strategy for now is to `sanitize()`/`cleanState()` (https://github.com/torch/nn/pull/526) and save the entire model via `torch.save`.

**Comment by gheinrich** [comment]
Hi @Liuftvafas I finally got around to doing this. I know it's late but can you review/try #701? Thanks!
